gobblin-core/src/main/java/gobblin/writer/AvroHdfsDatePartitionedWriter.java

```diff
@@ -0,0 +1,271 @@
+    this.properties = destination.getProperties();
+
+    // Confirm that all input parameters are not null
+    Preconditions.checkNotNull(this.destination);
+    Preconditions.checkNotNull(this.writerId);
+    Preconditions.checkNotNull(this.schema);
+    Preconditions.checkNotNull(this.writerOutputFormat);
+    Preconditions.checkNotNull(this.branch);
+    Preconditions.checkNotNull(this.properties);
+
+    this.baseFilePath = this.properties.getProp(getWriterFilePath(this.branch));
+
+    // Initialize the partitionLevel
+    this.partitionLevel =
+        this.properties.getProp(getWriterPartitionLevel(this.branch), ConfigurationKeys.DEFAULT_WRITER_PARTITION_LEVEL);
+
+    // Initialize the timestampToPathFormatter
+    this.timestampToPathFormatter =
+        DateTimeFormat.forPattern(this.properties.getProp(getWriterPartitionPattern(this.branch),
+            ConfigurationKeys.DEFAULT_WRITER_PARTITION_PATTERN));
+
+    // Check that ConfigurationKeys.WRITER_PARTITION_COLUMN_NAME has been specified and is properly formed
+
+    Preconditions.checkArgument(this.properties.contains(getWriterPartitionColumnName(this.branch)),
+        "Missing required property " + ConfigurationKeys.WRITER_PARTITION_COLUMN_NAME);
+
+    this.partitionColumnName = this.properties.getProp(getWriterPartitionColumnName(this.branch));
+    Optional<Schema> writerPartitionColumnSchema = AvroUtils.getFieldSchema(this.schema, this.partitionColumnName);
+
+    Preconditions.checkArgument(writerPartitionColumnSchema.isPresent(), "The column " + this.partitionColumnName
+        + " specified by " + ConfigurationKeys.WRITER_PARTITION_COLUMN_NAME + " is not in the writer input schema");
+
+    Preconditions.checkArgument(writerPartitionColumnSchema.get().getType().equals(Schema.Type.LONG), "The column "
+        + this.partitionColumnName + " specified by " + ConfigurationKeys.WRITER_PARTITION_COLUMN_NAME
+        + " must be of type " + Schema.Type.LONG);
+  }
+
+  @Override
+  public void write(GenericRecord record) throws IOException {
+
+    // Retrieve the value of the field specified by this.partitionColumnName
+    Optional<Object> writerPartitionColumnValue = AvroUtils.getFieldValue(record, this.partitionColumnName);
+    assert writerPartitionColumnValue.isPresent();
+
+    Path writerOutputPath = getPathForColumnValue((Long) writerPartitionColumnValue.get());
+
+    // If the path is not in pathToWriterMap, construct a new DataWriter, add it to the map, and write the record
+    // If the path is in pathToWriterMap simply retrieve the writer, and write the record
+    if (!this.pathToWriterMap.containsKey(writerOutputPath)) {
+
+      LOG.info("Creating a new DataWriter for path: " + new Path(this.baseFilePath, writerOutputPath));
+
+      DataWriter<GenericRecord> avroHdfsDataWriter = createAvroHdfsDataWriterForPath(writerOutputPath);
+
+      this.pathToWriterMap.put(writerOutputPath, avroHdfsDataWriter);
+      avroHdfsDataWriter.write(record);
```

Comment:
For some reason the schema compatibility check returns COMPATIBLE but writers still fail with ArrayOutOfBoundException for some schemas. So, at least for the Kafka case checking compatibility doesn't seem helpful. The current approach looks good.